### ISSUE TITLE
Simplify Materialization Scheduling

### ISSUE THREAD

**Issue by shangyian** [DataJunction/dj#809]
The current configure-a-materialization API is quite complex and can be simplified, since most of the metadata provided by users can be inferred from existing node metadata.

### Pre-Materialization Step

For nodes where we expect to materialize partition by partition, the user can set columns on the node to be temporal partition columns via their column attributes:
```
curl -X POST 'http://localhost:8000/nodes/default.repairs_cube/columns/default_DOT_total_repair_cost/attributes'
-d '{
    "namespace": "system",
    "name": "temporal_partition_day"
}'
```
We'll provide two new attribute tags for this purpose: `temporal partition day` and `temporal partition hour`. Setting these column attributes implies that any materialized version of the node should be partitioned by that column.

### Configure Materialization 

This means that the user wants DJ to schedule the appropriate materialization workflows in their scheduler of choice, based on the provided config. This will reuse the existing `POST /nodes/{node_name}/materializations` endpoint, but with some changes to the input schema.

User input that we'll keep at this stage:
- node to materialize
- schedule (cron)
- engine
- optional engine-specific config

We'll remove `partitions` from the expected input, since this can now be derived from column attributes. Additionally, if it's a Druid materialization, we'll remove the requirement for any Druid-specific config, as all of it can be derived at runtime based on column attributes.

DJ will store this materialization config and create a scheduled workflow and an adhoc workflow (which is used for kicking off backfills and adhoc runs). Note: If there are no temporal partitions, it will assume the node is meant to be materialized fully at the specified schedule. 

### Set Backfills

For a given node with a materialization configured, users can ask for the node to be backfilled (across some time range). This will trigger the adhoc workflow that was scheduled previously. This will be a new API endpoint:

```
curl -X POST 'http://localhost:8000/nodes/{node_name}/materializations/{materialization_name}/backfill'
-d '{
    "range": [
        "20230901",
        "20230930"
    ]
}'
```

If it's a Druid materialization, based on the configured partitions, it will generate the Druid config at runtime:
- we can assign granularity based on column attributes
- we can assign the timestamp_column based on column attributes
- intervals are generated based on the backfill partition (i.e., we should only load one interval at a time, whereas previously we were loading all intervals for each backfill job)

**Comment by shangyian** [comment]
@agorajek what do you think of the above? I know we don't have to change materialization atm, since it works, but I think it would be nice to have a UI for users to configure materialization in (especially if they want to build cubes), and it's better to create the UI after we clean up the API layer.

**Comment by agorajek** [comment]
@shangyian this sounds great! One question I have is how do we translate the `$dj_logical_timestamp` into the values of `temporal partition day` and `temporal partition hour` for the scheduled job, if the user did not make that translation in their SQL code?

**Comment by shangyian** [comment]
@agorajek good question... I think right now we do that in the query service (as in, we add an extra wrapper that adds in this where clause for the partition), but we should change that so that the core DJ service generates that query with the parameterized `$dj_logical_timestamp` and passes it to the query service for scheduling during the "configure materialization" phase.

**Comment by agorajek** [comment]
Right... so maybe instead of asking for  `temporal partition day|hour` lets ask the user to specify `temporal partition` and then `format` using some standard code (say Python), https://docs.python.org/3/library/datetime.html#format-codes ?

Remember that they already tell us the schedule, so we don't need to know the `... day|hour` part for anything other than backfills. But for backfills we can actually ask for that specifically as in:
1. Pick one partition column to run the backfill for.
2. 
    a. You picked a categorical partition, ok so please tell us the list of values to run the backfill for.
    b. You picked a temporal partition, ok so please tell us the time interval and the min, max values to run it with.

**Comment by shangyian** [comment]
That makes sense, except that then it's harder to use the existing column attributes for. I guess we could add some column-attribute-level metadata that stores the timestamp format.

**Comment by agorajek** [comment]
+1 on providing input for all partitions. And for the categorical it should be ok to skip, as-in no values set.

> That makes sense, except that then it's harder to use the existing column attributes for. I guess we could add some column-attribute-level metadata that stores the timestamp format.

Sorry for jumping around, but I just recalled this https://github.com/DataJunction/dj/discussions/421 and realized we could just ask for an expression when they define the temp partition, say: `CAST(FORMAT(${dj_logical_timestamp}, "YYYMMDD") AS INT)` and we can evaluate this expression to make sure it matches the type of the column. With this in hand asking for backfill range could be based on pure timestamp, cause we would know how to translate it into the partition values.
